### ISSUE TITLE
Improve implementation

### ISSUE THREAD

**Issue by ckaraneen** [ckarageorgkaneen/biornn#1]
Based on an exchange with Thomas Miconi:

> Hi,
> 
> Looking at outputs, it only learned to always output "Russian". That's because the dataset is massively imbalanced, at multiple scales. "Russian" names are about 49% of the whole dataset. Then English, etc.
> 
> I also noticed what seemed to be a couple of bugs? The recurrent layer seemed to be run multiple times, once for each element in weight_keys - with different perturbations. The loss function produced a single scalar rather than one value for each batch element. Etc.
> 
> A possible source of bugs is that the code is quite complex and perhaps over--engineered, with the logic dispersed over many interacting parts.
> 
> I put together a much simplified node perturbation version - with balanced sampling of all classes (I think). It's slow, but it kind of works. At 1200 epochs, it gets to ~.45 training accuracy, and almost .4 validation accuracy. (You can run it for fewer epochs to get a sense of how it does).
> 
> Thomas-

and his updated file notebook: [Updated_Ref_SGD_CutLR_BLnoperturb_biornn.zip](https://github.com/ckarageorgkaneen/biornn/files/12828893/Updated_Ref_SGD_CutLR_BLnoperturb_biornn.zip)
